### PR TITLE
rpc: add NEP11 and NNS interfaces to RPC Client

### LINKED ISSUE #1836: RPC client interface for NEP-11 and name service
We need a generic RPC client interface for non-fungible tokens as well as NNS-specific `resolve` and `isAvailable`.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #1836

### CODE REVIEW

File: pkg/rpc/server/server_test.go [hunk 1]
```diff
@@ -1540,12 +1540,12 @@ func testRPCProtocol(t *testing.T, doRPCCall func(string, string, *testing.T) []
...
+		t.Run("time frame only", func(t *testing.T) { testNEP17T(t, 4, 5, 0, 0, []int{8, 9, 10, 11}, []int{2, 3}) })
 		t.Run("no res", func(t *testing.T) { testNEP17T(t, 100, 100, 0, 0, []int{}, []int{}) })
-		t.Run("limit", func(t *testing.T) { testNEP17T(t, 1, 7, 3, 0, []int{3, 4}, []int{0}) })
-		t.Run("limit 2", func(t *testing.T) { testNEP17T(t, 4, 5, 2, 0, []int{6}, []int{1}) })
-		t.Run("limit with page", func(t *testing.T) { testNEP17T(t, 1, 7, 3, 1, []int{5, 6}, []int{1}) })
-		t.Run("limit with page 2", func(t *testing.T) { testNEP17T(t, 1, 7, 3, 2, []int{7, 8}, []int{2}) })
+		t.Run("limit", func(t *testing.T) { testNEP17T(t, 1, 7, 3, 0, []int{5, 6}, []int{1}) })
+		t.Run("limit 2", func(t *testing.T) { testNEP17T(t, 4, 5, 2, 0, []int{8}, []int{2}) })
+		t.Run("limit with page", func(t *testing.T) { testNEP17T(t, 1, 7, 3, 1, []int{7, 8}, []int{2}) })
+		t.Run("limit with page 2", func(t *testing.T) { testNEP17T(t, 1, 7, 3, 2, []int{9, 10}, []int{3}) })
```
Comment (roman-khimov) [inline pkg/rpc/server/server_test.go:1548]: Yeah, that's what write-only code looks like. I hope it's correct.

File: pkg/rpc/client/nep11.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+func (c *Client) NEP11TotalSupply(tokenHash util.Uint160) (int64, error) {
+	return c.nepTotalSupply(tokenHash)
+}
+
+// NEP11BalanceOf invokes `balanceOf` NEP11 method on a specified contract.
+func (c *Client) NEP11BalanceOf(tokenHash, owner util.Uint160) (int64, error) {
+	return c.nepBalanceOf(tokenHash, owner, nil)
+}
+
+// TODO: support NEP11TokensOf? (it returns iterator)
```
Comment (roman-khimov) [inline pkg/rpc/client/nep11.go:38]: It'd need some fancy script and paging support. Let's create a separate issue for that.

File: pkg/rpc/client/nep11.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+// Divisible NFT methods section end.
+
+// Optional NFT methods section start.
+
+// TODO: support Tokens? (it returns iterator)
+
+// NEP11PropertiesOptional invokes `properties` optional NEP11 method on a
+// specified contract.
+func (c *Client) NEP11PropertiesOptional(tokenHash util.Uint160, tokenID string) (*stackitem.Map, error) {
```
Comment (roman-khimov) [inline pkg/rpc/client/nep11.go:149]: `s/Optional//`, it's either supported or not, but the method name shouldn't be "optional".

File: pkg/rpc/client/nep11.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/rpc/client/nep11.go:38]: And other iteratorish methods.

File: pkg/rpc/client/nep11.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		},
+		Account: acc,
+	}})
+}
+
+// Non-divisible NFT methods section start.
+
+// NEP11OwnerOfNonDivisible invokes `ownerOf` non-devisible NEP11 method with the
+// specified token ID on a specified contract.
+func (c *Client) NEP11OwnerOfNonDivisible(tokenHash util.Uint160, tokenID string) (util.Uint160, error) {
```
Comment (roman-khimov) [inline pkg/rpc/client/nep11.go:89]: `NEP11NDOwnerOf`?

File: pkg/rpc/client/nep11.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+// Non-divisible NFT methods section end.
+
+// Divisible NFT methods section start.
+
+// TransferNEP11Divisible creates an invocation transaction that invokes 'transfer'
+// method on a given token to move specified amount of divisible NEP11 assets
+// (in FixedN format using contract's number of decimals) to given account and
+// sends it to the network returning just a hash of it.
+func (c *Client) TransferNEP11Divisible(acc *wallet.Account, to util.Uint160,
```
Comment (roman-khimov) [inline pkg/rpc/client/nep11.go:115]: `TransferNEP11D`?

File: pkg/rpc/client/nep11.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	return c.SendRawTransaction(tx)
+}
+
+// TODO: support NEP11OwnerOf? (it returns iterator)
+
+// NEP11BalanceOfDivisible invokes `balanceOf` divisible NEP11 method on a
+// specified contract.
+func (c *Client) NEP11BalanceOfDivisible(tokenHash, owner util.Uint160, tokenID string) (int64, error) {
```
Comment (roman-khimov) [inline pkg/rpc/client/nep11.go:137]: `NEP11DBalanceOf`? I think these names can be explained well and they're a little easier to type.

File: pkg/rpc/client/native.go [hunk 7]
```diff
@@ -63,3 +64,51 @@ func (c *Client) GetDesignatedByRole(role noderoles.Role, index uint32) (keys.Pu
 	}
 	return topPublicKeysFromStack(result.Stack)
 }
+
+// Resolve invokes `resolve` method on a native NameService contract.
+func (c *Client) Resolve(name string, typ recordtypes.Type) (string, error) {
```
Comment (roman-khimov) [inline pkg/rpc/client/native.go:69]: `NNSResolve`?

Comment (roman-khimov) [reply pkg/rpc/client/native.go:69]: And I'd filter out CNAME `typ` here.

File: pkg/rpc/client/native.go [hunk 8]
```diff
@@ -63,3 +64,51 @@ func (c *Client) GetDesignatedByRole(role noderoles.Role, index uint32) (keys.Pu
...
+	}
+	err = getInvocationError(result)
+	if err != nil {
+		return "", fmt.Errorf("`resolve`: %w", err)
+	}
+	return topStringFromStack(result.Stack)
+}
+
+// IsAvailable invokes `isAvailable` method on a native NameService contract.
+func (c *Client) IsAvailable(name string) (bool, error) {
```
Comment (roman-khimov) [inline pkg/rpc/client/native.go:95]: `NNSIsAvailable`?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/rpc/client/nep11.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/rpc/client/nep11.go:38]: https://github.com/nspcc-dev/neo-go/issues/1864

File: pkg/compiler/native_test.go [hunk 9]
```diff
@@ -77,10 +78,10 @@ func TestRoleManagementRole(t *testing.T) {
 }
 
 func TestNameServiceRecordType(t *testing.T) {
-	require.EqualValues(t, native.RecordTypeA, nameservice.TypeA)
-	require.EqualValues(t, native.RecordTypeCNAME, nameservice.TypeCNAME)
-	require.EqualValues(t, native.RecordTypeTXT, nameservice.TypeTXT)
-	require.EqualValues(t, native.RecordTypeAAAA, nameservice.TypeAAAA)
+	require.EqualValues(t, recordtypes.A, nameservice.TypeA)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/compiler/native_test.go:81]: https://github.com/nspcc-dev/neo-go/issues/1865

File: pkg/compiler/native_test.go [hunk 10]
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/nspcc-dev/neo-go/pkg/core/interop/interopnames"
 	"github.com/nspcc-dev/neo-go/pkg/core/native"
 	"github.com/nspcc-dev/neo-go/pkg/core/native/noderoles"
+	"github.com/nspcc-dev/neo-go/pkg/core/native/recordtypes"
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/compiler/native_test.go]: Sounds nice, I just couldn't choose the right name.

File: pkg/rpc/server/server_test.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/rpc/server/server_test.go:1548]: The logic is simple, I added two new transfers at the start of the `send` array and one new transfer at the start of the `received` array. The number of transfer is so large that it's far away from page 2, so page 2 stays the same and all indexes in this group of tests should just be moved to +2 for `send` and +1 for `received` accordingly.

File: pkg/rpc/client/native.go (hunk 7 above)
Comment (AnnaShaleva) [reply pkg/rpc/client/native.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**